### PR TITLE
Tiva CAN driver

### PR DESCRIPTION
**Pull request by MTres19**
## Summary
Char-device CAN driver for Tiva MCUs. The current implementation is blocked by #2663

## Impact
Minimal

## Testing
With DGTech Beacon CAN box

### CODE REVIEW

**Comment by acassis** [comment]
@MTres19 please fix these coding styles issues:
```
Error: /home/runner/work/incubator-nuttx/incubator-nuttx/nuttx/arch/arm/src/tiva/common/tiva_can.c:2166:49: error: Missing whitespace after semicolon
Error: /home/runner/work/incubator-nuttx/incubator-nuttx/nuttx/boards/arm/tiva/tm4c123g-launchpad/src/tm4c_bringup.c:63:1: error: Blank line contains whitespace
```
And please squash these commits in a single commit to let us to merge it.

File: arch/arm/src/tiva/common/tiva_can.c [hunk 1]
```diff
@@ -0,0 +1,2401 @@
...
+    .sf_id2   = 0,
+    .sf_type  = CAN_FILTER_MASK,
+    .sf_prio  = CAN_MSGPRIO_HIGH,
+  };
+#endif
+
+  /* REVISIT: This will not necessarily come out nicely on boards with
+   * different crystals.
+   */
+
```
Comment (hartmannathan) [inline arch/arm/src/tiva/common/tiva_can.c:464]: Can this be calculated from SYSCLK_FREQUENCY defined in the board's board.h?

File: arch/arm/src/tiva/hardware/tiva_can.h [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This is part of revision 2.2.0295 of the Tiva Peripheral Driver Library.
+ *
+ ****************************************************************************/
+
```
Comment (hartmannathan) [inline arch/arm/src/tiva/hardware/tiva_can.h:49]: Need to add this to LICENSE?

I think it's enough to add this filename, `arch/arm/src/tiva/hardware/tiva_can.h`, to the list that currently starts at line 1680 of LICENSE in the project root.

**Review by hartmannathan — COMMENTED** [review]
Wow, this was a big undertaking. I don't have a setup to test this PR on real hardware at the moment, but on code reading it looks pretty good. Thanks for commenting the code very well. When you're ready to take this out of "Draft" feel free to ping me...

**Comment by MTres19** [comment]
Thanks, I wanted to post it for review even though there's a couple of things still in progress. I'll squash and clean up for nxstyle when I have a chance.

For the bit-timing, SYSCLK_FREQUENCY actually is already used to calculate the baud rate prescaler. But it's still bad because TSEG1, TSEG2, and SJW are fixed and may not be correct when the division doesn't come out even. Ideally the application would never work with something as low-level as that because the optimal configuration depends on the clock frequency. I wrote a function (not specifically for NuttX, but I could try to see if I still have it) to determine the optimal configuration based on crystal tolerances between the different ECUs on the bus and the maximum values for TSEG1 and TSEG2 and SJW. It would be nice so implement something like that in the upper-half driver and just hook into the lower-half driver to get the available settings.

File: arch/arm/src/tiva/common/tiva_can.c (hunk 1 above)
Comment (MTres19) [reply arch/arm/src/tiva/common/tiva_can.c:464]: See comment: https://github.com/apache/incubator-nuttx/pull/5151#issuecomment-1008236696

**Comment by hartmannathan** [comment]
I'll try to review tomorrow morning. Thanks!

**Review by hartmannathan — CHANGES_REQUESTED** [review]
It seems that all TM4C Tiva chips that we currently support also have support for CAN1. I checked the datasheets for all of these:

TM4C123AH6PM
TM4C123GH6ZRB
TM4C123GH6PZ
TM4C123GH6PGE
TM4C1294NCPDT
TM4C129XNCZAD
TM4C129ENCZAD
TM4C129ENCPDT

This means we should either

(1) add "select TIVA_HAVE_CAN1" to all those chips, or
(2) add "select TIVA_HAVE_CAN1" to ARCH_CHIP_TM4C

If there are any TM4C for which there is no CAN1, then option (2) will mean that if architectural support is added for them, the "select TIVA_HAVE_CAN1" will need to move to individual chips, which we might forget to do...

I don't have a suitable hardware setup to test with, but on code reading it looks mostly good with just the few minor feedbacks mentioned inline...

Not related to this PR, but in arch/arm/src/tiva/Kconfig I noticed a doubling of the "config ARCH_CHIP_TM4C123GH6PGE" line which was already present in this file. One of us should probably fix that in a separate PR...

**Comment by MTres19** [comment]
Thanks for the feedback. Not all TM4C devices have dual CAN: https://www.ti.com/lit/sg/spmt285e/spmt285e.pdf

Should the changes to arch/arm/src/tiva/Kconfig go in this PR or a different one?

**Comment by hartmannathan** [comment]
> Thanks for the feedback. Not all TM4C devices have dual CAN: https://www.ti.com/lit/sg/spmt285e/spmt285e.pdf

Ah I see; many of the TM4C123 parts have only one CAN MAC. So in that case, I recommend to add `select TIVA_HAVE_CAN1` individually to TM4C123AH6PM, TM4C123GH6ZRB, TM4C123GH6PZ, TM4C123GH6PGE, TM4C1294NCPDT, TM4C129XNCZAD, TM4C129ENCZAD, and TM4C129ENCPDT, *not* to ARCH_CHIP_TM4C. This way, if/when architectural support is added for other TM4C12x parts, we will be able to specify the correct number of CAN MACs for the chip.

> Should the changes to arch/arm/src/tiva/Kconfig go in this PR or a different one?

Up to you... feel free to either make the changes or let me know that you'd like to have this PR merged as-is. I'm okay with either one.

**Comment by hartmannathan** [comment]
Reviewing......

**Review by hartmannathan — APPROVED** [review]
LGTM. I don't have actual hardware to test with other than that, this PR looks good. Thanks!

**Pull request merged**